PR: https://github.com/odoo/odoo/pull/98565

From: 23faf6f7ef966adaec37cfc6e83420c39c96f305
From: Raphael Collet
Date: 2023-09-19 16:37:03

Structural Changes: 5
Total Changes: 23

[FIX] *: compute methods mixing stored and non-stored fields

closes odoo/odoo#98565

Related: odoo/upgrade#5158
Related: odoo/enterprise#47508
Signed-off-by: Raphael Collet <rco@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_session.py
+++ b/addons/point_of_sale/models/pos_session.py
@@ -55,8 +55,8 @@ class PosSession(models.Model):
 
     opening_notes = fields.Text(string="Opening Notes")
     closing_notes = fields.Text(string="Closing Notes")
-    cash_control = fields.Boolean(compute='_compute_cash_all', string='Has Cash Control', compute_sudo=True)
-    cash_journal_id = fields.Many2one('account.journal', compute='_compute_cash_all', string='Cash Journal', store=True)
+    cash_control = fields.Boolean(compute='_compute_cash_control', string='Has Cash Control')
+    cash_journal_id = fields.Many2one('account.journal', compute='_compute_cash_journal', string='Cash Journal', store=True)
 
     cash_register_balance_end_real = fields.Monetary(
         string="Ending Balance",
@@ -154,15 +154,20 @@ class PosSession(models.Model):
         action['domain'] = [('id', 'in', self.picking_ids.ids)]
         return action
 
+    @api.depends('cash_journal_id')
+    def _compute_cash_control(self):
+        # Only one cash register is supported by point_of_sale.
+        for session in self:
+            if session.cash_journal_id:
+                session.cash_control = session.config_id.cash_control
+            else:
+                session.cash_control = False
+
     @api.depends('config_id', 'payment_method_ids')
-    def _compute_cash_all(self):
+    def _compute_cash_journal(self):
         # Only one cash register is supported by point_of_sale.
         for session in self:
-            session.cash_journal_id = session.cash_control = False
             cash_journal = session.payment_method_ids.filtered('is_cash_count')[:1].journal_id
-            if not cash_journal:
-                continue
-            session.cash_control = session.config_id.cash_control
             session.cash_journal_id = cash_journal
 
     @api.constrains('config_id')

--- a/addons/point_of_sale/models/res_config_settings.py
+++ b/addons/point_of_sale/models/res_config_settings.py
@@ -39,7 +39,6 @@ class ResConfigSettings(models.TransientModel):
     module_pos_preparation_display = fields.Boolean(string="Preparation Display", help="Show orders on the preparation display screen.")
     update_stock_quantities = fields.Selection(related="company_id.point_of_sale_update_stock_quantities", readonly=False)
     account_default_pos_receivable_account_id = fields.Many2one(string='Default Account Receivable (PoS)', related='company_id.account_default_pos_receivable_account_id', readonly=False)
-    is_default_pricelist_displayed = fields.Boolean(compute="_compute_pos_pricelist_id", compute_sudo=True)
     barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', related='company_id.nomenclature_id', readonly=False)
     is_kiosk_mode = fields.Boolean(string="Is Kiosk Mode", default=False)
 
@@ -273,9 +272,6 @@ class ResConfigSettings(models.TransientModel):
                     res_config.pos_available_pricelist_ids = res_config.pos_config_id.available_pricelist_ids
                     res_config.pos_pricelist_id = res_config.pos_config_id.pricelist_id
 
-            # TODO: Remove this field in master because it's always True.
-            res_config.is_default_pricelist_displayed = True
-
     @api.depends('pos_available_pricelist_ids', 'pos_use_pricelist')
     def _compute_pos_allowed_pricelist_ids(self):
         for res_config in self:
